xilem/src/view/virtual_scroll.rs

```diff
@@ -0,0 +1,355 @@
+        let widget = Pod::new(
+            widgets::VirtualScroll::<Element>::new(0).with_valid_range(self.valid_range.clone()),
+        );
+        ctx.record_action(widget.id);
+        (
+            widget,
+            private::VirtualScrollState {
+                pending_action: None,
+                previous_views: HashMap::default(),
+                current_updated: false,
+                pending_children_update: false,
+                current_views: HashMap::default(),
+                view_states: HashMap::default(),
+                my_path: ctx.view_path().into(),
+                cleanup_queue: Vec::default(),
+            },
+        )
+    }
+
+    fn rebuild(
+        &self,
+        prev: &Self,
+        view_state: &mut Self::ViewState,
+        ctx: &mut ViewCtx,
+        mut element: xilem_core::Mut<'_, Self::Element>,
+    ) {
+        if self.valid_range != prev.valid_range {
+            widgets::VirtualScroll::set_valid_range(&mut element, self.valid_range.clone());
+        }
+        if ctx.state_changed() && !view_state.pending_children_update {
+            let proxy = ctx.proxy();
+            proxy
+                .send_message(view_state.my_path.clone(), Box::new(UpdateVirtualChildren))
+                .unwrap();
+            view_state.pending_children_update = true;
+            // Either rebuilding or not rebuilding would be fine here. We choose not to
+            // because we know another rebuild is coming once this message is handled.
+        }
+        let used_action = view_state.pending_action.is_some();
+        if let Some(pending_action) = view_state.pending_action.take() {
+            debug_assert!(view_state.current_updated);
+            widgets::VirtualScroll::will_handle_action(&mut element, &pending_action);
+        }
+        if !view_state.current_updated {
+            // The set of children is the same, and our app state hasn't changed. Use a fast path for this.
+            for (idx, view) in &view_state.previous_views {
+                let state = view_state
+                    .view_states
+                    .get_mut(idx)
+                    .expect("All View states are accounted for");
+                if state.requested_rebuild {
+                    ctx.with_id(view_id_for_index(*idx), |ctx| {
+                        // Note that we rebuild the view with itself, because we're only actioning a requested rebuild.
+                        // This should be a no-op other than again the requested rebuild.
+                        view.rebuild(
+                            view,
+                            &mut state.state,
+                            ctx,
+                            widgets::VirtualScroll::child_mut(&mut element, *idx),
+                        );
+                    });
+                    state.requested_rebuild = false;
+                }
+            }
+        } else {
+            debug_assert!(view_state.cleanup_queue.is_empty());
```

Comment:
It may be useful to add a comment that this section is for when the set of children has changed.